### ISSUE TITLE
ICE: Assertion `width < BitWidth && "Invalid APInt Truncate request"' failed.

### ISSUE THREAD

**Issue by chengniansun** [llvm/llvm-project#50216]
|  |  |
| --- | --- |
| Bugzilla Link | [50872](https://llvm.org/bz50872) |
| Version | trunk |
| OS | Linux |
| CC | @DougGregor,@zygoloid |

## Extended Description 
$ clang-trunk -v
clang version 13.0.0 (https://github.com/llvm/llvm-project.git ef906573a127cffef7cae75d5155c15a8a2a3a5e)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/cnsun/usr/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/10
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/10
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64

$ cat mutant.c
typedef double float64x1_t __attribute__((vector_size(8))) float64x1_t arg1 =
    (float64x1_t)0x3fedf9d4343c7c80;

$ clang-trunk  mutant.c
mutant.c:1:59: error: expected ';' after top level declarator
typedef double float64x1_t __attribute__((vector_size(8))) float64x1_t arg1 =
                                                          ^
                                                          ;
clang-13: /tmp/tmp.B1HwadcXIH-clang-builder/llvm-project/llvm/lib/Support/APInt.cpp:863: llvm::APInt llvm::APInt::trunc(unsigned int) const: Assertion `width < BitWidth && "Invalid APInt Truncate request"' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /scratch/software/clang-trunk/bin/clang-13 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all --mrelax-relocations -disable-free -main-file-name mutant.c -mrelocation-model static -mframe-pointer=all -fmath-errno -fno-rounding-math -mconstructor-aliases -munwind-tables -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fcoverage-compilation-dir=/scratch/cnsun/workspace/perses-fuzzer/default_c_finding_folder/crash_20210625_091111_0df0/delta/perses_result -resource-dir /scratch/software/clang-trunk/lib/clang/13.0.0 -c-isystem . -c-isystem /usr/local/include/cmsith -internal-isystem /scratch/software/clang-trunk/lib/clang/13.0.0/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/10/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdebug-compilation-dir=/scratch/cnsun/workspace/perses-fuzzer/default_c_finding_folder/crash_20210625_091111_0df0/delta/perses_result -ferror-limit 19 -fgnuc-version=4.2.1 -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/mutant-e9a893.o -x c mutant.c
1.	mutant.c:2:36: current parser token ';'
 #&#8203;0 0x000055a3d9ed11f4 PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #&#8203;1 0x000055a3d9ece9ae SignalHandler(int) Signals.cpp:0:0
 #&#8203;2 0x00007fe22fb433c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x153c0)
 #&#8203;3 0x00007fe22f5e218b raise /build/glibc-ZN95T4/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #&#8203;4 0x00007fe22f5c1859 abort /build/glibc-ZN95T4/glibc-2.31/stdlib/abort.c:81:7
 #&#8203;5 0x00007fe22f5c1729 get_sysdep_segment_value /build/glibc-ZN95T4/glibc-2.31/intl/loadmsgcat.c:509:8
 #&#8203;6 0x00007fe22f5c1729 _nl_load_domain /build/glibc-ZN95T4/glibc-2.31/intl/loadmsgcat.c:970:34
 #&#8203;7 0x00007fe22f5d2f36 (/lib/x86_64-linux-gnu/libc.so.6+0x36f36)
 #&#8203;8 0x000055a3d9df0d02 llvm::APInt::trunc(unsigned int) const (/scratch/software/clang-trunk/bin/clang-13+0x3878d02)
 #&#8203;9 0x000055a3dd01fc64 (anonymous namespace)::VectorExprEvaluator::VisitCastExpr(clang::CastExpr const*) ExprConstant.cpp:0:0
#&#8203;10 0x000055a3dd021701 clang::StmtVisitorBase<llvm::make_const_ptr, (anonymous namespace)::VectorExprEvaluator, bool>::Visit(clang::Stmt const*) ExprConstant.cpp:0:0
#&#8203;11 0x000055a3dd022361 EvaluateVector(clang::Expr const*, clang::APValue&, (anonymous namespace)::EvalInfo&) ExprConstant.cpp:0:0
#&#8203;12 0x000055a3dcfc89e6 Evaluate(clang::APValue&, (anonymous namespace)::EvalInfo&, clang::Expr const*) ExprConstant.cpp:0:0
#&#8203;13 0x000055a3dcfd2c89 EvaluateAsRValue((anonymous namespace)::EvalInfo&, clang::Expr const*, clang::APValue&) ExprConstant.cpp:0:0
#&#8203;14 0x000055a3dcfd4d4f clang::Expr::EvaluateAsRValue(clang::Expr::EvalResult&, clang::ASTContext const&, bool) const (/scratch/software/clang-trunk/bin/clang-13+0x6a5cd4f)
#&#8203;15 0x000055a3dcfd4dd4 clang::Expr::isEvaluatable(clang::ASTContext const&, clang::Expr::SideEffectsKind) const (/scratch/software/clang-trunk/bin/clang-13+0x6a5cdd4)
#&#8203;16 0x000055a3dcf95100 clang::Expr::isConstantInitializer(clang::ASTContext&, bool, clang::Expr const**) const (/scratch/software/clang-trunk/bin/clang-13+0x6a1d100)
#&#8203;17 0x000055a3dc36cd54 clang::Sema::CheckForConstantInitializer(clang::Expr*, clang::QualType) (/scratch/software/clang-trunk/bin/clang-13+0x5df4d54)
#&#8203;18 0x000055a3dc3b4874 clang::Sema::AddInitializerToDecl(clang::Decl*, clang::Expr*, bool) (/scratch/software/clang-trunk/bin/clang-13+0x5e3c874)
#&#8203;19 0x000055a3dc10a5c6 clang::Parser::ParseDeclarationAfterDeclaratorAndAttributes(clang::Declarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::ForRangeInit*) (/scratch/software/clang-trunk/bin/clang-13+0x5b925c6)
#&#8203;20 0x000055a3dc11d4ab clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::SourceLocation*, clang::Parser::ForRangeInit*) (/scratch/software/clang-trunk/bin/clang-13+0x5ba54ab)
#&#8203;21 0x000055a3dc0eb5b6 clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/scratch/software/clang-trunk/bin/clang-13+0x5b735b6)
#&#8203;22 0x000055a3dc0ebba1 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (.part.0) Parser.cpp:0:0
#&#8203;23 0x000055a3dc0f23a7 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) (/scratch/software/clang-trunk/bin/clang-13+0x5b7a3a7)
#&#8203;24 0x000055a3dc0f393e clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, bool) (/scratch/software/clang-trunk/bin/clang-13+0x5b7b93e)
#&#8203;25 0x000055a3dc0e6189 clang::ParseAST(clang::Sema&, bool, bool) (/scratch/software/clang-trunk/bin/clang-13+0x5b6e189)
#&#8203;26 0x000055a3db02d458 clang::CodeGenAction::ExecuteAction() (/scratch/software/clang-trunk/bin/clang-13+0x4ab5458)
#&#8203;27 0x000055a3da8dc9e9 clang::FrontendAction::Execute() (/scratch/software/clang-trunk/bin/clang-13+0x43649e9)
#&#8203;28 0x000055a3da86fc36 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/scratch/software/clang-trunk/bin/clang-13+0x42f7c36)
#&#8203;29 0x000055a3da9ba870 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/scratch/software/clang-trunk/bin/clang-13+0x4442870)
#&#8203;30 0x000055a3d77da696 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/scratch/software/clang-trunk/bin/clang-13+0x1262696)
#&#8203;31 0x000055a3d77d69d8 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#&#8203;32 0x000055a3d77040e6 main (/scratch/software/clang-trunk/bin/clang-13+0x118c0e6)
#&#8203;33 0x00007fe22f5c30b3 __libc_start_main /build/glibc-ZN95T4/glibc-2.31/csu/../csu/libc-start.c:342:3
#&#8203;34 0x000055a3d77d654e _start (/scratch/software/clang-trunk/bin/clang-13+0x125e54e)
clang-13: error: unable to execute command: Aborted (core dumped)
clang-13: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 13.0.0 (https://github.com/llvm/llvm-project.git ef906573a127cffef7cae75d5155c15a8a2a3a5e)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/cnsun/usr/bin
clang-13: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-13: note: diagnostic msg: /tmp/mutant-154909.c
clang-13: note: diagnostic msg: /tmp/mutant-154909.sh
clang-13: note: diagnostic msg:

********************

**Comment by AaronBallman** [comment]
You can also reproduce the crash on valid code:
```
typedef double float64x1_t __attribute__((vector_size(8)));
float64x1_t arg1 = (float64x1_t)0x3fedf9d4343c7c80;
```

**Comment by AaronBallman** [comment]
This was fixed in 859d92394812c0e5e650ab337717719b9687e5ce
